appengine/genetics/js/blocks.js

```diff
@@ -0,0 +1,465 @@
+   * @this Blockly.Block
+   */
+  init: function() {
+    this.jsonInit({
+      "message0": "%1.%2",
+
+      "args0": [
+        {
+          "type": "input_value",
+          "name": "MOUSE",
+          "check": "Mouse"
+        },
+        {
+          "type": "field_dropdown",
+          "name": "PROPERTY",
+          "options": [
+            ['size', 'SIZE'],
+            ['aggressiveness', 'AGGRESSIVENESS'],
+            ['startAggressiveness', 'START_AGGRESSIVENESS'],
+            ['fertility', 'FERTILITY'],
+            ['startingFertility', 'START_FERTILITY'],
+            ['sex', 'SEX'],
+            ['age', 'AGE'],
+            ['id', 'ID'],
+            ['pickFightOwner', 'PICK_FIGHT'],
+            ['chooseMateOwner', 'CHOOSE_MATE'],
+            ['mateAnswerOwner', 'MATE_ANSWER']
+          ]
+        }
+      ],
+      "inputsInline": true,
+      "output": "Number",
+      "colour": Genetics.Blocks.GENETICS_HUE
+    });
+    // Assign 'this' to a variable for use in the tooltip closure below.
+    var thisBlock = this;
+    this.setTooltip(function() {
+      var mode = thisBlock.getFieldValue('PROPERTY');
+      var TOOLTIPS = {
+        'SIZE': BlocklyGames.getMsg('Genetics_sizeTooltip'),
+        'AGGRESSIVENESS': BlocklyGames.getMsg('Genetics_aggressivenessTooltip'),
+        'FERTILITY': BlocklyGames.getMsg('Genetics_fertilityTooltip'),
+        'START_FERTILITY': BlocklyGames.getMsg('Genetics_startFertilityTooltip'),
+        'SEX': BlocklyGames.getMsg('Genetics_sexTooltip'),
+        'AGE': BlocklyGames.getMsg('Genetics_ageTooltip'),
+        'ID': BlocklyGames.getMsg('Genetics_idTooltip'),
+        'PICK_FIGHT': BlocklyGames.getMsg('Genetics_pickFightOwnerTooltip'),
+        'CHOOSE_MATE': BlocklyGames.getMsg('Genetics_chooseMateOwnerTooltip'),
+        'MATE_ANSWER': BlocklyGames.getMsg('Genetics_mateAnswerOwnerTooltip')
+      };
+      return TOOLTIPS[mode];
+    });
+  },
+  /**
+   * Modify this block to have the correct output type.
+   * @param {string} newProp Either a property requesting a function owner that
+   * returns a string or a property that returns a number.
+   * @private
+   * @this Blockly.Block
+   */
+  updateType_: function(newProp) {
+    if (newProp == 'SEX' || newProp == 'PICK_FIGHT' ||
+        newProp == 'CHOOSE_MATE' || newProp == 'MATE_ANSWER') {
+      this.outputConnection.setCheck('String');
+    } else {
+      this.outputConnection.setCheck('Number');
+    }
+  }
+};
+
+/**
+ * Defines the JavaScript generation for getProperties.
+ * @param {Blockly.Block} block
+ * @return {!Array.<string|number>}
+ */
+Blockly.JavaScript['genetics_getProperties'] = function(block) {
+  // Generate JavaScript for getting mouse property.
+  var mouse = Blockly.JavaScript.valueToCode(block, 'MOUSE',
+          Blockly.JavaScript.ORDER_NONE) || 'me()';
+  var property = block.getFieldValue('PROPERTY');
+  var code = mouse + '.';
+  switch (property) {
+    case 'SIZE':
+      code += 'size';
```

Comment:
This seems to be repeating a bunch of stuff from the options in lines 267 onwards.

Is it possible to define the mapping in some common JS object and refer to that object both from here and from line 267?